### ISSUE TITLE
A connection slips after the server is shut down

### ISSUE THREAD

**Issue by srogatch** [oatpp/oatpp#269]
I launch `Server::run()` in one thread, then in another thread I stop the server as follows:
```cpp
pConnectionProvider->close();
pConnectionHandler->stop();
pServer->stop();
thrMain.join(); // thrMain is the one running Server::run()
```
However, I've noticed a connection slipping through to a `HttpRequestHandler::handle()` after the above code is executed.
Is the above code correct for stopping the server? Does the server exit `run()` method before all `HttpRequestHandler::handle()` have finished? If so, does `oatpp` provide a way to wait till all the `handle()` methods finish after stopping the server?
The platform is Windows 10 64-bit, the compiler is Visual Studio 2019 v16.6.2 . The oatpp version is 1.1.0 .

**Comment by lganzzzo** [comment]
Hello @srogatch ,

## The Problem

I have investigated the issue - the problem is that the connection is managed by an already allocated thread and the connection is treated as `keep-alive`.
Meaning that despite the fact that server is already stopped and connection provider is not listening on port anymore, you still have an open connection in the separate thread that is serving requests.

Please consider the following example program:

```cpp
#include "oatpp/web/server/HttpConnectionHandler.hpp"

#include "oatpp/network/server/Server.hpp"
#include "oatpp/network/server/SimpleTCPConnectionProvider.hpp"

#include <thread>

class Handler : public oatpp::web::server::HttpRequestHandler {
private:
  std::shared_ptr<oatpp::network::ServerConnectionProvider> m_connectionProvider;
  std::shared_ptr<oatpp::network::server::ConnectionHandler> m_connectionHandler;
  oatpp::network::server::Server* m_server;
public:

  Handler(const std::shared_ptr<oatpp::network::ServerConnectionProvider>& connectionProvider,
          const std::shared_ptr<oatpp::network::server::ConnectionHandler>& connectionHandler,
          oatpp::network::server::Server* server)
    : m_connectionProvider(connectionProvider)
    , m_connectionHandler(connectionHandler)
    , m_server(server)
  {}

  std::shared_ptr<OutgoingResponse> handle(const std::shared_ptr<IncomingRequest>& request) override {

    //*****************************
    // Stop server on first request

    m_server->stop();
    m_connectionProvider->close();
    m_connectionHandler->stop();

    //*****************************

    return ResponseFactory::createResponse(Status::CODE_200, "Hello World!");

  }

};

void run() {

  auto router = oatpp::web::server::HttpRouter::createShared();

  auto connectionHandler = oatpp::web::server::HttpConnectionHandler::createShared(router);
  auto connectionProvider = oatpp::network::server::SimpleTCPConnectionProvider::createShared(8000 /* port */);
  oatpp::network::server::Server server(connectionProvider, connectionHandler);

  /* Route GET - "/hello" requests to Handler */
  router->route("GET", "/hello", std::make_shared<Handler>(connectionProvider, connectionHandler, &server));

  OATPP_LOGI("MyApp", "Server running on port %s", connectionProvider->getProperty("port").getData());

  server.run(); //<----- Run server

  OATPP_LOGI("MyApp", "Server stopped!");

  std::this_thread::sleep_for(std::chrono::minutes(10));

}

int main() {
  oatpp::base::Environment::init();
  run();
  oatpp::base::Environment::destroy();
  return 0;
}
```

This program will stop the server immediately on the first request to http://localhost:8000/hello endpoint.
So if we do two separate requests:

```bash
$ curl http://localhost:8000/hello #<--- the first one is successful 
Hello World!

$ curl http://localhost:8000/hello #<--- the second one fails
curl: (7) Failed to connect to localhost port 8000: Connection refused
```

However, if you keep your connection alive - you can still serve requests through that connection:

```bash
$ ab -k -c 1 -n 100000 http://localhost:8000/hello
...
Benchmarking localhost (be patient)
Completed 10000 requests
Completed 20000 requests
Completed 30000 requests
Completed 40000 requests
Completed 50000 requests
Completed 60000 requests
Completed 70000 requests
Completed 80000 requests
Completed 90000 requests
Completed 100000 requests
Finished 100000 requests
...
Concurrency Level:      1
Time taken for tests:   5.236 seconds
Complete requests:      100000
Failed requests:        0
Keep-Alive requests:    100000
```

All `100000` requests were served!

## The Solution

You may just put `Connection: Close` header to the response and the allocated thread will be closed immediately after sending the response to the client.

```cpp
auto response = ResponseFactory::createResponse(Status::CODE_200, "Hello World!");
response->putHeader(Header::CONNECTION, Header::Value::CONNECTION_CLOSE);
return response;
```

**Comment by lganzzzo** [comment]
It's probably a bug,  and it might be a good idea to check if the `ConnectionHandler` is stopped so that the worker thread would exit.

**Comment by srogatch** [comment]
Thanks for your investigation, @lganzzzo .  However, it may be that what you have observed is a slightly different problem. My problem is that I need a join point where it is guaranteed that the object derived from `HttpRequestHandler` will no more receive a `handle()` request (and all previous `handle()` requests have exited). I believe that `response->putHeader(Header::CONNECTION, Header::Value::CONNECTION_CLOSE);` would negatively impact the performance for clients who send multiple requests and it still doesn't guarantee that no thread is in `handle()` by the moment the server is stopped.
So your another suggestion to check if the `ConnectionHandler` has stopped seems to fit my needs better. But how to do this? It seems that `oatpp::web::server::HttpConnectionHandler` has only methods `stop()` (which I'm calling) and `handleConnection()`, which is obviously unrelated.

**Comment by srogatch** [comment]
So far a workaround I use is:
```
class Handler : public oatpp::web::server::HttpRequestHandler {
  bool _isShutDown = false;
  std::shared_mutex _mu;
public:
  std::shared_ptr<OutgoingResponse> handle(const std::shared_ptr<IncomingRequest>& request) override {
    std::shared_lock<std::shared_mutex> ml(_mu);
    if(_isShutDown) {
      throw std::runtime_error("A connection came after the handler is shut down.");
    }
    return ResponseFactory::createResponse(Status::CODE_200, "Hello World!");
  }
  void Shutdown() {
    std::unique_lock<std::shared_mutex> ml(_mu);
    _isShutDown = true;
  }
};
```
Then in the thread that shuts everything down, I do:
```
pConnectionProvider->close();
pConnectionHandler->stop();
pServer->stop();
pHandler->Shutdown();
thrMain.join(); // thrMain is the one running Server::run()
```
If you don't have a better solution, can we move the shared mutex to `oatpp` layer so to really prevent `handle()` calls after we call `pConnectionHandler->stop()`?

Please, note that `HttpConnectionHandler::stop()` declaration has the following comment near `HttpConnectionHandler.hpp:99` :
```
  /**
   * Tell all worker threads to exit when done.
   */
  void stop() override;
```
That looks like an obligation to prevent `handle()` calls after this point.

**Comment by lganzzzo** [comment]
Hello @srogatch ,

> So your another suggestion to check if the ConnectionHandler has stopped seems to fit my needs better. But how to do this? It seems that oatpp::web::server::HttpConnectionHandler has only methods stop() (which I'm calling) and handleConnection(), which is obviously unrelated.

The idea I currently have - is to share one `atomic<bool> isRunning` between all threads spawned by the `ConnectionHandler`, and to check it's value on per-request basis here - https://github.com/oatpp/oatpp/blob/master/src/oatpp/web/server/HttpProcessor.cpp#L198

> If you don't have a better solution, can we move the shared mutex to oatpp layer so to really prevent handle() calls after we call pConnectionHandler->stop()?

Unfortunately, we can't use `shared_mutex` at the moment as oatpp min-supported C++ version is 11. the `shared_mutex` requires C++ 17.

---

The difference in the proposed solutions is:
Consider the situation when you do some very long operation in the `handle` method (it may be a client request to a remote resource)
- shared_mutex approach - stop-calling thread will sleep on `pHandler->Shutdown();` until that long-running handle is finished.
- atomic in HttpProcessor task - stop-calling thread will return immediately on `pHandler->Shutdown();` but the long-running handle will continue to run until it finished.

---

I believe that `shared_mutex` solution is more consistent. But again it's C++ 17 :(.
Also, I'm not sure about the performance impact of the `shared_mutex` can introduce.

Regards,
Leonid

**Comment by srogatch** [comment]
With that `atomic`, if the `Shutdown()` method always returns immediately, how would you know that all the long-running `handle()` operations have finished? My intent was to have a `thread::join()`-like solution that blocks until all handlers have finished.

There are lightweight platform-dependent replacements for a `shared_mutex`, e.g. https://docs.microsoft.com/en-us/windows/win32/sync/slim-reader-writer--srw--locks

**Comment by lganzzzo** [comment]
Hey @srogatch 

> My intent was to have a thread::join()-like solution that blocks until all handlers have finished.

I agree - such behavior is consistent and is exactly what user would expect.
Also, I don't really see much sense in returning immediately and have handler still running.

**My proposition for now would be:** to fork oatpp and implement the std::shared_mutex approach on the oatpp level. I'll review the code if needed.

As for the main oatpp repo, for now - this issue will remain open and unresolved - as I don't like existing options, neither the atomic, nor the platform dependent solutions.

Regards,
Leonid

**Comment by lganzzzo** [comment]
Hey @srogatch ,

Did you manage to achieve the consistent server stoppage functionality, or you decided to go with the proposed above workaround (shared mutex on the handlers' level)?

Please let me know if you need help from my side.

Regards,
Leonid

**Comment by srogatch** [comment]
Hi @lganzzzo ! I'm currently using the `std::shared_mutex` as I showed above. I took a look at the code, but it seems that to extend `oatpp` a deep understanding of the source code is needed but currently I cannot spend that much time on it. Thanks for offering your help. Maybe after our release, I will have more time for this issue.

**Comment by lganzzzo** [comment]
Fixed in `1.3.0`. See [proper-server-stoppage](https://github.com/oatpp/oatpp/blob/master/changelog/1.3.0.md#proper-server-stoppage)
